discover/utils.go

```diff
@@ -0,0 +1,177 @@
+package discover
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/roasbeef/btcd/btcec"
+	"github.com/roasbeef/btcd/wire"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+// getNameByMsg retrieves string representation if announce message type.
+func getNameByMsg(msg interface{}) string {
+	return reflect.TypeOf(msg).String()
+}
+
+// getInterface takes type as input and return its interface. This function
+// mainly used in type switch cases.
+func getInterface(t reflect.Type) interface{} {
+	return reflect.Zero(t).Interface()
+}
+
+// createHandlerFunc takes the interface as input and checks that interface is
+// corresponding to the handler function with specific schema, if schema is
+// wrong or input isn't a function it panics.
+func createHandlerFunc(f interface{}) (string, Handler) {
```

Comment:
In my opinion, this package currently over-uses reflection. Instead, (as I mentioned above) I think a more concrete interface should be used, with further type assertions to obtain concrete types usable depending on the transport.